### PR TITLE
mrc-4177 save dependencies when saving a task

### PR DESCRIPTION
**Pull request by hillalex**
Save `depends_on` info alongside a task when saving a new task. Things we might also want to do here:

* initialise dependent tasks with a different status 
* verify that dependencies are satisfied on `task_run`
* mark tasks as impossible when dependencies are cancelled

But with the way logic is distributed between this, `queuer` and `didehpc`, it's a little unclear who should be responsible for this sort of thing, and there's an argument for leaving all of the above to the other packages, who are actually orchestrating the tasks.

### CODE REVIEW

File: R/task_bulk.R [hunk 1]
```diff
@@ -38,7 +40,7 @@ bulk_task_save <- function(X, FUN, context, DOTS = NULL,
   db <- context$db
   context_id <- context$id
 
-  verify_dependencies_exist(depends_on, context)
+  verify_dependencies_exist(unlist(depends_on), context)
```
Comment (richfitz) [inline R/task_bulk.R]: Looks like this verification should go after the validation (lines 55-61)?

File: R/task_bulk.R [hunk 2]
```diff
@@ -50,6 +52,13 @@ bulk_task_save <- function(X, FUN, context, DOTS = NULL,
   }
 
   n <- length(dat)
+  if (!is.null(depends_on)) {
+    nd <- length(depends_on)
+    if (nd < n) {
+      stop(sprintf(paste("Failed to save as 'depends_on' must be of",
```
Comment (richfitz) [inline R/task_bulk.R]: do you want to support the shorthand of `depends_on` being a single vector shared across everything? so if `depends_on` is a character vector of ids then you'd do `depends_on <- rep(list(depends_on), n)`?

Comment (hillalex) [reply R/task_bulk.R]: Yeah I was wondering whether that might be worth doing. But might be a bit ambiguous if the vector is the same length as the list of tasks?

Comment (richfitz) [reply R/task_bulk.R]: it's a different type though in that case (a list vs a character vector), which should make it less ambiguous

**Review by richfitz — APPROVED** [review]

**Pull request merged**